Thread

flagarde opened the pull request:
This is an other PR for input. @TobiasWallner I think this should improve performances.

TobiasWallner commented:
I am currently in my hollidays, will have a look in 1 or 2 weeks. 

Always nice to hear from your advancements.

TobiasWallner commented on cpp-terminal/platforms/input.cpp:
```diff
@@ -26,19 +45,167 @@ std::string to_utf8(LPCWCH utf16Str)
+  {
+#if defined(_WIN32)
+    Term::Event ret;
+    WaitForSingleObject(Term::Private::in.handle(), INFINITE);
```
Oh, nice so now we have an extra thread that actually brings the OS to wait for some input instead of polling every 10ns, right?

flagarde replied:
Yes, and we ask the OS to hang until there is an event. I guess this would solve very fast typing who could be merged as one event.

TobiasWallner replied:
I have not yet used the c++ std threads a lot. When we pause one thread with a breaking system call, does only this thread pause or does the whole process pause? Do you have any information on that?

flagarde replied:
I have changed the events.cpp in example. I add a thread printing some text to see if the thread is stopped. In linux and windows it seems to continue.

TobiasWallner commented on examples/slow_events.cpp:
```diff
@@ -0,0 +1,73 @@
+#include <thread>
+
+// Same than event but add 1s to slow down the event processing.
+int main()
```
do we also have an equivalent fast_events.cpp?

flagarde replied:
The events.cpp should be the "fast" default one.. the slow is just to stress the fact that key pressed could be merged into one event due to the loop time se had before.

TobiasWallner reviewed (approved):
nice done. I am currently working on some changes on my application. Cannot wait to try it out. I will probably feel a little sharper especially when typing fast.

flagarde commented on cpp-terminal/event.cpp:
```diff
@@ -310,10 +310,7 @@ void Term::Event::parse(const std::string& str)
       m_container.m_Key = Key(Term::Key::Value::Numeric5);
     else if(str.size() == 2 && ((str[0] & 0b11100000) == 0b11000000) && ((str[1] & 0b11000000) == 0b10000000)) { m_container.m_Key = Key(static_cast<Term::Key::Value>(Term::Private::utf8_to_utf32(str)[0])); }
     else if(str.size() == 3 && ((str[0] & 0b11110000) == 0b11100000) && ((str[1] & 0b11000000) == 0b10000000) && ((str[2] & 0b11000000) == 0b10000000)) { m_container.m_Key = Key(static_cast<Term::Key::Value>(Term::Private::utf8_to_utf32(str)[0])); }
-    else if(str.size() == 4 && ((str[0] & 0b11111000) == 0b11110000) && ((str[1] & 0b11000000) == 0b10000000) && ((str[2] & 0b11000000) == 0b10000000) && ((str[3] & 0b11000000) == 0b10000000))
```
Yes, need to change that and move to hex

flagarde replied:
@TobiasWallner This has been fixed mainstream as well as many Windows warnings

flagarde commented:
Before merging other PR I would like to merge this one as it is a really big change in the way cpp-terminal is processing the events. I hope it will solve some problems. Have you tried it with Lime ?

flagarde commented:
I found some strane bugs on macos so I need moe time to fix

flagarde commented:
@TobiasWallner @certik I'm more confident now with this PR, I have cleaned many things and added workaround for the strange bugs in macos. I have try with Lime but maybe you could have a try yourself because you knoe it better than me ^^. Still maybe some cleaning but I think it is clean anough to be merged
